strawberry/types/types.py

```diff
@@ -23,26 +33,257 @@
 
+T = TypeVar("T")
+
+
+def _get_interfaces(cls: Type[T]) -> List[StrawberryDefinition]:
+    interfaces = []
+
+    for base in cls.__bases__:
+        type_definition = cast(
+            Optional[StrawberryDefinition],
+            getattr(base, "__strawberry_definition__", None),
+        )
+
+        if type_definition and type_definition.is_interface:
+            interfaces.append(type_definition)
+
+        for inherited_interface in _get_interfaces(base):
+            interfaces.append(inherited_interface)
+
+    return interfaces
+
+
+def _get_fields(cls: Type) -> List["StrawberryField"]:
+    """Get all the strawberry fields off a strawberry.type cls
+
+    This function returns a list of StrawberryFields (one for each field item), while
+    also paying attention the name and typing of the field.
+
+    StrawberryFields can be defined on a strawberry.type class as either a dataclass-
+    style field or using strawberry.field as a decorator.
+
+    >>> import strawberry
+    >>> @strawberry.type
+    ... class Query:
+    ...     type_1a: int = 5
+    ...     type_1b: int = strawberry.field(...)
+    ...     type_1c: int = strawberry.field(resolver=...)
+    ...
+    ...     @strawberry.field
+    ...     def type_2(self) -> int:
+    ...         ...
+
+    Type #1:
+        A pure dataclass-style field. Will not have a StrawberryField; one will need to
+        be created in this function. Type annotation is required.
+
+    Type #2:
+        A field defined using @strawberry.field as a decorator around the resolver. The
+        resolver must be type-annotated.
+
+    The StrawberryField.python_name value will be assigned to the field's name on the
+    class if one is not set by either using an explicit strawberry.field(name=...) or by
+    passing a named function (i.e. not an anonymous lambda) to strawberry.field
+    (typically as a decorator).
+    """
+    # Deferred import to avoid import cycles
+    from strawberry.field import StrawberryField
+
+    fields: Dict[str, StrawberryField] = {}
+
+    # TODO: What is this?
+    # Find the class that each field was originally defined in, so we can use
+    # that scope later when resolving the type, as it may have different names
+    # available to it.
+    origins: Dict[str, type] = {field_name: cls for field_name in cls.__annotations__}
+
+    for base in cls.__mro__:
+        if __strawberry_definition__ := getattr(
+            base, "__strawberry_definition__", False
+        ):
+            # TODO: remove this when merge with StrawberryObject
+            cast(__strawberry_definition__, StrawberryDefinition)
+            for field in __strawberry_definition__.fields:
+                if field.python_name in base.__annotations__:
+                    origins.setdefault(field.python_name, base)
+
+    return list(fields.values())
+
 
 @dataclasses.dataclass(eq=False)
-class TypeDefinition(StrawberryType):
+class StrawberryDefinition(StrawberryType):
```

Comment:
1. @patrick91, what do you think of this renaming. I wonder if calling this `StrawberryDefinition` is too broad and this should be `StrawberryObjectDefinition` or `StrawberryTypeDefinition`
2. @nrbnlulu, I hate to ask this because it's going to turn into a fair bit of work for you but can you please do this renaming in a separate diff? This turns a large diff into an absolutely massive diff making this fields change really hard to review.

    This renaming of `TypeDefinition` -> `StrawberryDefinition` is also a somewhat breaking change -- we want to separate it from the fields change to allow easier version upgrading by our users

Reply:
> Also, I have rename `_type_defenition` to `__strawberry_definition__` I hope you will forgive me not putting this in another PR, though `_type_defenition` is kinda cryptic :upside_down_face: ...

Just saw this comment :grimacing:. I think splitting it out is pretty important, unfortunately

Reply:
doing it in another PR sounds best :) this will break strawberry django and other libs I think :)

for the name I'd go with StrawberryTypeDefinition for object types, StrawberryInputTypeDefinition for input types and StrawberryInterfaceDefinition for interfaces. But that makes me think of how we have StrawberryScalar, which is not too consistent with these names, maybe we can rename it to StrawberryScalarDefinition (something for another PR )

Reply:
O.K. although we can have a property with a warning that will return a pointer to the new name...